### PR TITLE
Port to fabric 1.15.2, initial feedback pass

### PR DESCRIPTION
**Pull request by Sturmlilie**
*Do not merge, this PR is against the 1.14 branch!*

Hi! I took it upon myself to port the fabric branch to 1.15.2. This is my first time doing anything with MC modding / fabric / Java. The branch is not yet in a state to be merged, but it's very playable at this point and so I wanted to get an initial round of feedback before I progress. The biggest changes in 1.15 were rendering related, and so this is also what most of my code touches. I've made incremental changes in order for every commit (except the first few ones) to build and load, so if I replaced a certain interface, often I left the old methods in tact for the code to still build (eg: `RenderUtil:drawQuad`); eventually those old interfaces will be removed when their last user is gone / converted.

At this point every module renders (albeit with some caveats), and the core game mechanics are intact. There were a couple apparent behavior bugs (~like copying ROM from RAM modules~) that I haven't touched yet because they also appear in the 1.14 branch this is based on.

## Things that are explicitly broken by me:
* In general, lighting is a bit off; I try to ignore skylight whereever possible, but the default `RenderLayers` still apply the legacy OpenGL lighting so there is a slight variance between eg. a module on top of a casing vs on one of the sides.
* ~All module text is darkened by a factor of 0.5 due to an issue I haven't yet been able to identify. I know it goes away when I turn off legacy OpenGL lighting (`glDisable(GL_LIGHTING)`), which worst case I can inject a custom cutout entity renderlayer with lighting turned off, but that would require a bunch of invasive Mixins and I'd rather try to get to the bottom of the issue first.~
* For some modules, I wasn't able to 100% replicate their original appearance due to problems with rendering transparent quads in the new architecture (more on that below).

## Regarding module overlay drawing
In the past this was done via immediate OpenGL and so the programmer had maximum flexibility. With the new render pipeline, submitted quads are not rendered on the spot and their ordering is also not defined, so instead of the painter's algorithm we have to either not overlap quads or rely on the depth buffer. I was able to mostly work around this by only drawing opaque quads and slightly bumping certain ones within the 0.005 Z-axis margin from the casing block (see [CasingBlockEntityRenderer](https://github.com/MightyPirates/TIS-3D/blob/master-MC1.14-fabric/src/main/java/li/cil/tis3d/client/render/block/entity/CasingBlockEntityRenderer.java#L108)).  
Now, if one wanted to keep that flexibility (which I wouldn't advocate for), all of the module overlay rendering would have to be moved to a dynamic texture, ie. every module get's their own X by X space on an FBO and draws their state in whatever fashion onto it, and the casing renderer then uses that to texture itself.  
However, seeing how I was able to reproduce 90% of the modules with the existing approach, I don't think it's worth the hassle.

Note that most of the methods added by me lack javadoc comments; if there are no bigger rewrites necessary, I will comment as many as possible in a followup commit.

Please feel free to critique any part of my code, I'm trying to learn =)

### CODE REVIEW

**Comment by fnuecke** [comment]
This is awesome! Thanks so much, I really appreciate it :)

I'll browse through the changes, but don't expect me to be able to tell you anything when it comes to 1.15, I expect I'll learn a lot from the diffs myself.

I did give it a quick start, and looks good indeed!

Responding to the dynamic overlay drawing, it was really nice to have while it worked, but I totally agree that it's not a must-have and everything that goes toward better render performance has my vote anyway, so that's cool. ~~Will be interesting to see if that z offset may need more bumping if we get people with a variety of graphics cards to use it; in my experience there's a chance for a bit of variance on different systems~~ Nevermind, seems to be same offset used in the casing renderer, so should be fine. I suggest making the one with the bump a separate method, so callers can decide for themselves if they need it, i.e. if the caller is sure the things on the same layer don't overlap (e.g. border outline and inner parts)?

ROM/RAM copying - hmm, to make sure we're talking about the same thing: copying the contents of a RAM module on a casing into a ROM module held in hand? That seems to work for me though (sneak rclick = copy to hand, regular rclick = copy from hand). In your branch. Did you fix that since? Am I missing something?

**Comment by Sturmlilie** [comment]
> Nevermind, seems to be same offset used in the casing renderer, so should be fine. I suggest making the one with the bump a separate method, so callers can decide for themselves if they need it, i.e. if the caller is sure the things on the same layer don't overlap (e.g. border outline and inner parts)?

The 0.005 bump by the casing renderer was there before and I highly suspect it's because otherwise the module overlay contents would Z-fight with the casing block. Most of the 1.14 code turned off the depth buffer very liberally, something that is also not available to us anymore. The bumps I do are on the order of 0.001, which I'm can't really confirm is safe for all video cards. I'll dig more into it (we might get Z-fighting when the casing is far away).

Edit: I just tested it by removing the 0.005 bump in the casing renderer, and it indeed ends up in z-fight club for every module, even the ones with just a single sprite. All in all it's a tradeof between guaranteed z-fight avoidance and having the overlay levitate too far above the casing block.

> ROM/RAM copying - hmm, to make sure we're talking about the same thing: copying the contents of a RAM module on a casing into a ROM module held in hand? That seems to work for me though (sneak rclick = copy to hand, regular rclick = copy from hand). In your branch. Did you fix that since? Am I missing something?

Apologies, I probably misread the instructions then, I thought just right clicking copied the contents from the installed RAM to the held ROM; it's probably not busted then ^^"

**Comment by fnuecke** [comment]
Offset: ah, all right then, thanks for testing it!

ROM/RAM: great, one bug fewer to worry about ;)

**Comment by Sturmlilie** [comment]
Managed to disable diffuse lighting for text / display module without involving Mixins at all :)

image

**Comment by fnuecke** [comment]
Awesome!

**Comment by Sturmlilie** [comment]
So, I went through all my commits and made a list of smaller fixes that I'll apply for the next round; if you have any requested changes, those will also go into the next round. I would kindly ask you to create a new 1.15 branch based off `master-MC1.14-fabric`, then I'll point my PR at it.

> I suggest making the one with the bump a separate method, so callers can decide for themselves if they need it

I hope I was able to address this in my earlier comment; basically it wouldn't make sense to make a separate method because every module needs the bump by default; any additional layers go inside that existing `0.005` margin.

**Comment by Sturmlilie** [comment]
Changes since opening the PR:
* Use mipmapped render layer whenever possible
* Use `final` more
* Fix many instances of wrong ARGB component order
* Restore original TerminalModule look
* Use the Z-margin equally when rendering multiple overlay layers (eg. 2nd layer at `0.0025` instead of `0.001`)
* Remove comment about ROM module being busted
* Reordered / split some commits up
* define `RenderUtil.maxLight` via `LightmaptextureManager` instead of magic value
* less apparent magic values in ControllerBlock entity label draw method
* Remove `CasingBlockEntityRenderer:AXIS_*` constants in favor of vanilly `Vector3f` constants
* Try to use lowercase `f` float suffix whereever possible

**Comment by Sturmlilie** [comment]
Excuse the churn, I forgot to fix one issue in one the very first commits and that causes all followups to change their hashes during rebase >.>

**Comment by fnuecke** [comment]
Hey! Sorry for the radio silence there; I'll try to make time to have a look through everything and do a bit of testing this week -- are there any known issues left or are you happy with the PR as it is now?

**Comment by Sturmlilie** [comment]
@fnuecke I'm happy with the PR as is. I'd like to reinstate the culling of invisible module overlays, but after 1.15 support is in mainline.

Oh, and javadoc comments I'd like to add once you're happy with the code as-is, I hope that's okay. One more thing I just remembered, I never got around to also exposing the new methods I added in `AbstractFontRenderer` in `FontRendererAPI`. I'm not quite sure I understand the inheritance tree here; `FontRendererAPI` is supposed to expose `SmallFontRenderer`, correct? Which would make implementing something like the Terminal Module from the outside impossible, since it relies on `NormalFontRenderer`.

File: src/main/java/li/cil/tis3d/api/prefab/manual/ItemStackTabIconRenderer.java [hunk 1]
```diff
@@ -25,8 +25,8 @@ public ItemStackTabIconRenderer(final ItemStack stack) {
     public void render() {
         GlStateManager.enableRescaleNormal();
         RenderUtil.ignoreLighting();
-        GuiLighting.enableForItems();
+        //~ DiffuseLighting.enableForItems();
```
Comment (fnuecke) [inline src/main/java/li/cil/tis3d/api/prefab/manual/ItemStackTabIconRenderer.java:28]: Is this open for investigation, or does this not have any visible effect? Either a todo comment or remove it.

File: src/main/java/li/cil/tis3d/client/manual/segment/render/ItemStackImageRenderer.java [hunk 2]
```diff
@@ -38,8 +38,8 @@ public void render(final int mouseX, final int mouseY) {
         GlStateManager.scalef(getWidth() / 16f, getHeight() / 16f, getWidth() / 16f);
         GlStateManager.enableRescaleNormal();
         RenderUtil.ignoreLighting();
-        GuiLighting.enableForItems();
+        //~ DiffuseLighting.enableForItems();
```
Comment (fnuecke) [inline src/main/java/li/cil/tis3d/client/manual/segment/render/ItemStackImageRenderer.java:41]: Same as with tab icon renderer -> todo comment or remove.

File: src/main/java/li/cil/tis3d/api/util/RenderUtil.java [hunk 3]
```diff
@@ -1,20 +1,49 @@
...
+    public static final int maxLight = LightmapTextureManager.pack(0xF, 0xF);
+    private static Sprite whiteSprite;
+
+    // Rationale: There is no standard RenderLayer that has all the standard
+    // elements but without texturing; 'Leash' has no blending and normals,
+    // 'Lightning' does have blending but with a blendfunc that isn't 'over'.
+    // Fabric has no method for creating custom RenderLayers, so the easiest
+    // and least likely to break method is to just use a white dummy texture
+    // for colored quads, plus there's less context switches due to the batching
+    // with regular textured quads.
```
Comment (fnuecke) [inline src/main/java/li/cil/tis3d/api/util/RenderUtil.java:38]: If there's no reason for this not to be a Javadoc style comment, change that?

Also maybe reword the  last section, since you added the `RenderLayerAccess` class (which effectively does add a layer) this can be a bit misleading. I do agree with this being better this way due to better batching.

Comment (fnuecke) [inline src/main/java/li/cil/tis3d/api/util/RenderUtil.java:38]: Getting a warning about missing sidedness annotation here. Probably via [Minecraft Development plugin](https://plugins.jetbrains.com/plugin/8327-minecraft-development). Needs an `@Environment(EnvType.CLIENT)` (due to client side `getSprite`).

File: src/main/java/li/cil/tis3d/api/util/RenderUtil.java [hunk 4]
```diff
@@ -148,14 +235,106 @@ public static void drawQuad(final Sprite sprite) {
...
+                                final int argb, final int l, final int ol) {
+
+        drawQuad(matrices, vc, x, y, w, h, u0, v0, u1, v1,
+                 ColorUtils.getAlphaU8(argb),
+                 ColorUtils.getRedU8(argb),
+                 ColorUtils.getGreenU8(argb),
+                 ColorUtils.getBlueU8(argb),
+                 l, ol);
+    }
+
```
Comment (fnuecke) [inline src/main/java/li/cil/tis3d/api/util/RenderUtil.java:293]: Missed a `@Environment(EnvType.CLIENT)` here.

File: src/main/java/li/cil/tis3d/common/block/CasingBlock.java [hunk 5]
```diff
@@ -149,26 +151,26 @@ public boolean activate(final BlockState state, final World world, final BlockPo
                             }
                         }
                     }
-                    return true;
+                    return ActionResult.SUCCESS; // XXX
```
Comment (fnuecke) [inline src/main/java/li/cil/tis3d/common/block/CasingBlock.java:154]: Why the `XXX`? (Here and below)

File: src/main/java/li/cil/tis3d/common/block/ControllerBlock.java [hunk 6]
```diff
@@ -50,7 +51,7 @@ public boolean activate(final BlockState state, final World world, final BlockPo
                         player.dropItem(bookManual, false, false);
                     }
                 }
-                return true;
+                return ActionResult.SUCCESS; // XXX
```
Comment (fnuecke) [inline src/main/java/li/cil/tis3d/common/block/ControllerBlock.java:54]: Why the XXX? (Here and below)

File: src/main/java/li/cil/tis3d/common/entity/InfraredPacketEntity.java [hunk 7]
```diff
@@ -370,14 +361,14 @@ private void onBlockCollision(final BlockHitResult hit) {
...
-        if (!world.isBlockLoaded(pos)) {
+        if (!WorldUtils.isBlockLoaded(world, pos)) {
             return;
         }
         final Block block = world.getBlockState(pos).getBlock();
 
         // Traveling through a portal?
         if (block == Blocks.NETHER_PORTAL || block == Blocks.END_PORTAL) {
-            setInPortal(pos);
+            setInNetherPortal(pos);
```
Comment (fnuecke) [inline src/main/java/li/cil/tis3d/common/entity/InfraredPacketEntity.java:371]: Is this just unfortunate mappings or is there a separate method `setInEndPortal`?

File: src/main/java/li/cil/tis3d/common/module/DisplayModule.java [hunk 8]
```diff
@@ -150,7 +158,7 @@ public void onDisposed() {
 
     @Override
     public void finalize() {
-        LeakDetector.add(glTextureId, nativeImage);
+        //~ LeakDetector.add(glTextureId, nativeImage);
```
Comment (fnuecke) [inline src/main/java/li/cil/tis3d/common/module/DisplayModule.java:161]: Does `NativeImageBackedTexture` mean the LeakDetector isn't needed anymore at all? If so, kill it. Otherwise rewrite to track those and close them instead of freeing the GL id instead?

File: src/main/java/li/cil/tis3d/common/module/AudioModule.java [hunk 9]
```diff
@@ -53,16 +57,19 @@ public void step() {
...
-    public void render(final BlockEntityRenderDispatcher rendererDispatcher, final float partialTicks) {
+    public void render(final BlockEntityRenderDispatcher rendererDispatcher, final float partialTicks,
+                       final MatrixStack matrices, final VertexConsumerProvider vcp,
+                       final int light, final int overlay) {
         if (!getCasing().isEnabled()) {
             return;
         }
 
-        GlStateManager.enableBlend();
+        // XXX Somemthing isn't quite right yet, the sprite seems bigger than in 1.14
```
Comment (fnuecke) [inline src/main/java/li/cil/tis3d/common/module/AudioModule.java:67]: Had a look side-by-side with 1.14 and looks pretty same to me?

File: src/main/java/li/cil/tis3d/common/module/SequencerModule.java [hunk 10]
```diff
@@ -124,46 +129,44 @@ public void onData(final ByteBuf data) {
...
-        GlStateManager.depthMask(false);
+        matrices.push();
+        rotateForRendering(matrices);
 
+        final VertexConsumer vcCutout = vcp.getBuffer(RenderLayer.getCutoutMipped());
+        final RenderLayer translLayer = RenderLayer.getTranslucent();
+        final VertexConsumer vcTransl = vcp.getBuffer(translLayer);
+        final MatrixStack.Entry mat = matrices.peek();
         final boolean enabled = getCasing().isEnabled();
-        if (enabled) {
```
Comment (fnuecke) [inline src/main/java/li/cil/tis3d/common/module/SequencerModule.java:142]: Wondering if there's a reason for the change of order here?

File: src/main/resources/fabric.mod.json [hunk 11]
```diff
@@ -3,7 +3,7 @@
     "id": "tis3d",
     "name": "TIS-3D",
     "description": "TIS-100 inspired low-tech computing for Minecraft.",
-    "version": "1.5.0",
+    "version": "1.5.99",
```
Comment (fnuecke) [inline src/main/resources/fabric.mod.json:6]: Since it's been that way before: the MC version was always kinda treated as a super-major version, so when targeting different MC versions the patch version was always restarted at zero. So unless it's for ease of development mind changing that back to 0?

File: src/main/java/li/cil/tis3d/common/init/Entities.java [hunk 12]
```diff
@@ -18,6 +20,9 @@
 
     static void registerEntities() {
         Registry.register(Registry.ENTITY_TYPE, Constants.NAME_ENTITY_INFRARED_PACKET, INFRARED_PACKET);
+
+        EntityRendererRegistry.INSTANCE.register(INFRARED_PACKET,
```
Comment (fnuecke) [inline src/main/java/li/cil/tis3d/common/init/Entities.java:24]: Should probably go to `BootstrapClient`? Crashes dedicated server on startup.

Side question: since it's rendering nothing, I'm guessing it's needed so MC doesn't complain that there's no renderer for the entity?

File: src/main/java/li/cil/tis3d/common/module/DisplayModule.java [hunk 13]
```diff
@@ -66,9 +73,12 @@ public static void tick() {
      * current state to newly connected/coming closer clients.
      */
     private final int[] image = new int[RESOLUTION * RESOLUTION];
-    private Object nativeImage;
     private boolean imageDirty = false;
 
+    private NativeImageBackedTexture backingTexture;
```
Comment (fnuecke) [inline src/main/java/li/cil/tis3d/common/module/DisplayModule.java:78]: Client class references break dedicated server. Includes this field and methods below used in rendering. Throwing `@Environment(EnvType.CLIENT)` at them should work.

**Review by fnuecke — COMMENTED** [review]
I made a note on one or two places where I got sided annotation warnings, but later I stumbled across a bunch more; too lazy to mark that here, so I'll just add them after the merge.

The Queue/Stack refactor was a bit big (and surprising, since not directly related to the port). I guess it makes sense. Only thing I'm not super happy with is the new closure in the render logic, but I can't think of better (while still readable) way. Unless Java is smarter than I give it credit for and that doesn't allocate each time, then never mind.

Question: I noticed the `GlStateManager` methods are all flagged as deprecated now but still used in UI classes. Do you know if the UI code is still "supposed" to be using that, or is there a new system for that it would make sense to switch over to, long term?


Again, thank you so much for the work!

Edit: oh right, re font rendering: yeah, the API is just for the small font renderer; was originally just intended for text such as on the execution and stack modules.

File: src/main/java/li/cil/tis3d/api/prefab/manual/ItemStackTabIconRenderer.java (hunk 1 above)
Comment (Sturmlilie) [reply src/main/java/li/cil/tis3d/api/prefab/manual/ItemStackTabIconRenderer.java:28]: Sorry for that. I didn't notice any weirdness in the GUI rendering, but now that I'm more comfortable with the code bases and after comparing the source of 1.14's `GuiLighting`, im 95% certain the new equivalent is `DiffuseLighting.enableGuiDepthLighting`.

File: src/main/java/li/cil/tis3d/api/util/RenderUtil.java (hunk 3 above)
Comment (Sturmlilie) [reply src/main/java/li/cil/tis3d/api/util/RenderUtil.java:38]: Yeah, at the time of the comment I was completely lost on how to define custom render layers; and now that I've managed to do it I still think this approach is okay. Will reformat the comment.

File: src/main/java/li/cil/tis3d/common/block/CasingBlock.java (hunk 5 above)
Comment (Sturmlilie) [reply src/main/java/li/cil/tis3d/common/block/CasingBlock.java:154]: Sorry about that, the `XXX` is because I didn't know which of the 4 `ActionResult`s is appropriate here and guessed. From yarn:
```
/**
    * Indicates an action is performed and the actor's hand should swing to
    * indicate the performance.
    */
   SUCCESS,
   /**
    * Indicates an action is performed but no animation should accompany the
    * performance.
    */
   CONSUME,
   /**
    * Indicates an action is not performed but allows other actions to
    * perform.
    */
   PASS,
   /**
    * Indicates that an action is not performed and prevents other actions
    * from performing.
    */
   FAIL;
```

What I should have done is observe the behavior / hand motion in 1.14 and make sure it is replicated (will do that).

File: src/main/java/li/cil/tis3d/common/entity/InfraredPacketEntity.java (hunk 7 above)
Comment (Sturmlilie) [reply src/main/java/li/cil/tis3d/common/entity/InfraredPacketEntity.java:371]: Likely the former. I don't remember touching this myself so it must have been the remapper; indeed 1.14's `setInPortal` and 1.15's `setInNetherPortal` look identical source-wise. Overall it seems that all references/names in `Entity` to "portal" have been rescoped to just the nether portal.

File: src/main/java/li/cil/tis3d/common/module/DisplayModule.java (hunk 8 above)
Comment (Sturmlilie) [reply src/main/java/li/cil/tis3d/common/module/DisplayModule.java:161]: I reviewed #93, but the issue opener never reported back so I was unsure what to do here. Personally I didn't encounter any crashes when testing the TextureManager unregister code, but that was only within single player. I will reinstate the leak detector code.

File: src/main/java/li/cil/tis3d/common/module/AudioModule.java (hunk 9 above)
Comment (Sturmlilie) [reply src/main/java/li/cil/tis3d/common/module/AudioModule.java:67]: I'll try to provide screenshots as proof that I'm (hopefully) not hallucinating.

File: src/main/java/li/cil/tis3d/common/module/SequencerModule.java (hunk 10 above)
Comment (Sturmlilie) [reply src/main/java/li/cil/tis3d/common/module/SequencerModule.java:142]: It's because of the relative `matrices.translate()` calls to push the current Z position forward, for the render order issue I discussed. I'll try to see if I can rephrase the render code to produce less changes though.  
However as I mentioned in the comment above, the ported visual appearance doesn't look good regardless; transparency is my biggest problem. I don't know why but the render layers offering blending always make the quads turn solid when viewed at an angle not close to 90° (you can see this yourself by moving up very close to the module and watching the rectangle pass by). This is probably lighting-related, I should try using a custom render layer with blending and no lighting, now that I know how to do it.

File: src/main/resources/fabric.mod.json (hunk 11 above)
Comment (Sturmlilie) [reply src/main/resources/fabric.mod.json:6]: Sure, the 99 was just for me to differentiate my WIP code, but I should have done that via `BUILD_NUMBER` instead of abusing the mod version. If the MC version is the super-major, that means the breaking changes in `api` are also covered by it, right?

File: src/main/java/li/cil/tis3d/common/init/Entities.java (hunk 12 above)
Comment (Sturmlilie) [reply src/main/java/li/cil/tis3d/common/init/Entities.java:24]: Yes, my bad.

> Side question: since it's rendering nothing, I'm guessing it's needed so MC doesn't complain that there's no renderer for the entity?

Correct, the client aborts at runtime if it doesn't find a registered renderer.

**Comment by Sturmlilie** [comment]
> I made a note on one or two places where I got sided annotation warnings, but later I stumbled across a bunch more; too lazy to mark that here, so I'll just add them after the merge.

Hmm, unfortunate that the gradle build task doesn't give me those warnings, I'll see if I can do something about that to save you the nerves.

> The Queue/Stack refactor was a bit big (and surprising, since not directly related to the port). I guess it makes sense. Only thing I'm not super happy with is the new closure in the render logic, but I can't think of better (while still readable) way. Unless Java is smarter than I give it credit for and that doesn't allocate each time, then never mind.

Yeah, and I totally expected pushback on it :) After porting the Stack Module, I looked at the Queue Module code and felt weird copy pasting all the rendering code, though it did make me nervous that I ended up having to touch a lot more parts of the classes for the subclassing to work. For the iteration code, I considered a bunch of approaches, first just naively having overridden `begin`/`end` methods that return an int index (didn't work due to the queue wrap-around), then using a dedicated `Iterator` implementation (lots of boilerplate for just one use-site), and settled on the closure due to its simplicity. It certainly allocates once to capture the outer variables, but after I would hope it only invokes the passed lambda without additional overhead.

But I have no problems with reversing the subclassing and just porting the modules each individually, if you're more comfortable with that.

> Question: I noticed the GlStateManager methods are all flagged as deprecated now but still used in UI classes. Do you know if the UI code is still "supposed" to be using that, or is there a new system for that it would make sense to switch over to, long term?

From a quick look through 1.15's sources, it appears that most `GlStateManager` methods marked as deprecated have been superseded by more general methods in `RenderSystem`. 1.15's internal GUIs also appear to use exclusively `Tesselator`/`BufferBuilder` (basically, immediate mode rendering that is as graphics API agnostic as possible) instead of making raw immediate mode GL calls; in fact 1.14 already did so too, albeit still relying on `GlStateManager`.

I haven't yet looked at 1.16 in detail but I know that all GUI classes get passed an additional `MatrixStack` for rendering now. I don't know if the `Tesselator` code is going to stay, but long term any move away from explicit OpenGL is probably a good thing.

Edit: I can try my hand at rewriting the TIS-3D GUI classes to be more future-proof if you'd like, although I'd like to get this mod ready for the 1.16 release first if possible.

> Again, thank you so much for the work!

Thanks for taking the time to review, and sorry about not testing this on a dedicated server, I naively assumed that since single player runs on an integrated server instance, it would cover all dedicated paths as well.

Actually I'm surprised to not have gotten push back on a couple other places, eg. the move to vector classes for the infrared packet positioning code.

> Edit: oh right, re font rendering: yeah, the API is just for the small font renderer; was originally just intended for text such as on the execution and stack modules.

I figured as much. Maybe the API should expose a method to select one of the different implementations? I feel kind of bad just leaving it broken, but it's up to you.

File: src/main/java/li/cil/tis3d/common/module/AudioModule.java (hunk 9 above)
Comment (Sturmlilie) [reply src/main/java/li/cil/tis3d/common/module/AudioModule.java:67]: 1.14 first, then my 1.15 version; it's barely noticable at the very edge of the block: https://youtu.be/gMSZ73vMJvQ

File: src/main/java/li/cil/tis3d/common/module/DisplayModule.java (hunk 8 above)
Comment (Sturmlilie) [reply src/main/java/li/cil/tis3d/common/module/DisplayModule.java:161]: Taking a closer look at this, I'm wondering why `deleteTexture()` was ever called from `finalize()` in the first place. The call from `onDisposed()` is correct assuming this method is called at least once near the end of the module's lifetime. Since the GC can in theory run on a different thread, `finalize()` is most definitely the wrong place for it.

I believe the `LeakDetector` is superfluous but I will leave its logic untouched since this is technically not related to the port.

File: src/main/java/li/cil/tis3d/common/block/CasingBlock.java (hunk 5 above)
Comment (Sturmlilie) [reply src/main/java/li/cil/tis3d/common/block/CasingBlock.java:154]: Alright that was foolish, `SUCCESS` is clearly the only return value that replicates the hand swinging.

**Comment by Sturmlilie** [comment]
Interesting optical illusion I encountered while working on the SequencerModule rendering:  
image

The innermost square in the bottom left is brighter than the one at the top.

**Comment by Sturmlilie** [comment]
Continuing in 2nd PR to avoid clutter: #101

File: src/main/java/li/cil/tis3d/common/module/AudioModule.java (hunk 9 above)
Comment (fnuecke) [reply src/main/java/li/cil/tis3d/common/module/AudioModule.java:67]: Weird, I'll have to check that apparently I've been hallucinating, because I could've sworn I didn't see that bleed into the outer "pixel" area, haha.

File: src/main/java/li/cil/tis3d/common/module/SequencerModule.java (hunk 10 above)
Comment (fnuecke) [reply src/main/java/li/cil/tis3d/common/module/SequencerModule.java:142]: OK, cool. Thanks for the explanation!

File: src/main/resources/fabric.mod.json (hunk 11 above)
Comment (fnuecke) [reply src/main/resources/fabric.mod.json:6]: Yes, those changes would be covered by that; pretty sure there's precedent for that, too.

File: src/main/java/li/cil/tis3d/common/init/Entities.java (hunk 12 above)
Comment (fnuecke) [reply src/main/java/li/cil/tis3d/common/init/Entities.java:24]: OK, cool :)
